src/core/event-handler.js

```diff
@@ -0,0 +1,221 @@
+
+                    this._callbackActive[key] = this._callbackActive[key].slice();
+                }
+            }
+
+            if (!name) {
+                this._callbacks = { };
+            } else if (!callback) {
+                if (this._callbacks[name])
+                    this._callbacks[name] = [];
+            } else {
+                var events = this._callbacks[name];
+                if (!events)
+                    return this;
+
+                var count = events.length;
+
+                for (var i = 0; i < count; i++) {
+                    if (events[i].callback !== callback)
+                        continue;
+
+                    if (scope && events[i].scope !== scope)
+                        continue;
+
+                    events[i--] = events[--count];
+                }
+                events.length = count;
+            }
+
+            return this;
+        },
+
+        // ESLint rule disabled here as documenting arg1, arg2...argN as [...] rest
+        // arguments is preferable to documenting each one individually.
+        /* eslint-disable valid-jsdoc */
+        /**
+         * @function
+         * @name pc.EventHandler#fire
+         * @description Fire an event, all additional arguments are passed on to the event listener
+         * @param {Object} name Name of event to fire
+         * @param {*} [arg1] First argument that is passed to the event handler
+         * @param {*} [arg2] Second argument that is passed to the event handler
+         * @param {*} [arg3] Third argument that is passed to the event handler
+         * @param {*} [arg4] Fourth argument that is passed to the event handler
+         * @param {*} [arg5] Fifth argument that is passed to the event handler
+         * @param {*} [arg6] Sixth argument that is passed to the event handler
+         * @param {*} [arg7] Seventh argument that is passed to the event handler
+         * @param {*} [arg8] Eighth argument that is passed to the event handler
+         * @returns {*} 'this' for chaining
+         * @example
+         * obj.fire('test', 'This is the message');
+         */
+        /* eslint-enable valid-jsdoc */
+        fire: function (name, arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8) {
+            if (!name || !this._callbacks[name])
+                return this;
+
+            var callbacks;
+
+            if (!this._callbackActive[name]) {
+                this._callbackActive[name] = this._callbacks[name];
+            } else {
+                if (this._callbackActive[name] === this._callbacks[name])
+                    this._callbackActive[name] = this._callbackActive[name].slice();
+
+                callbacks = this._callbacks[name].slice();
+            }
+
+            // TODO: What does callbacks do here?
+            // In particular this condition check looks wrong: (i < (callbacks || this._callbackActive[name]).length)
+            // Because callbacks is not an integer
```

Comment:
integer->array?

Reply:
I'm not sure! I've merely refactored the code in events.js to a base class format. The function contents, including the comments, are identical to what they were before.